### PR TITLE
Tempelis: GitOps for Slack

### PR DESCRIPTION
**Pull request by Katharine**
This PR adds Tempelis, which reads a file or directory of configuration files and effects that change against a running Slack instance. It can run in "dry run" mode (the default) in which it just prints out what it would do, or it can actually do things.

Also see [the design doc](https://docs.google.com/document/d/1SUEzTD9BUmj5bXHtQ4yCtjWjqa1AXoCKOrR8Eb8VDwQ/edit) (requires kubernetes-dev).

This is a draft because a) I have literally not tested it at all, and also b) it could maybe use some unit tests

/cc @BenTheElder

### CODE REVIEW

File: tempelis/reconciler/reconciler.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	a, e := r.reconcileChannels()
+	actions = append(actions, a...)
+	errors = append(errors, e...)
+	a, e = r.reconcileUsergroups()
+	actions = append(actions, a...)
+	errors = append(errors, e...)
+
+	failed := false
+	if len(errors) > 0 {
+		log.Printf("This configuration cannot be applied against the current reality:")
```
Comment (BenTheElder) [inline tempelis/reconciler/reconciler.go:61]: this is a great log message.

**Review by BenTheElder — COMMENTED** [review]
generally LGTM, will come back for a second pass when it's passing presubmits 😅 

before we deploy this we should probably add tests

we need a logo to go with the greek name 😛

File: slack/types.go [hunk 2]
```diff
@@ -57,15 +57,61 @@ type User struct {
...
+	IsPrivate          bool          `json:"is_private"`
+	IsMPIM             bool          `json:"is_mpim"`
+	LastRead           string        `json:"last_read"`
+	Topic              struct {
+		Topic   string `json:"value"`
+		Creator string `json:"ID"`
+		LastSet int64  `json:"last_set"`
+	} `json:"topic"`
+	Purpose struct {
+		Purpose string `json:"value"`
```
Comment (BenTheElder) [inline slack/types.go:110]: Conversation.Purpose.Purpose = value? /Shrug

**Review by BenTheElder — APPROVED** [review]
Tentatively LGTM, but this is a lot.
If you'd rather merge and iterate go ahead, otherwise sometime tomorrow I'm happy to take another pass over it.
/lgtm
/hold

File: slack/types.go (hunk 2 above)
Comment (Katharine) [reply slack/types.go:110]: Slack calls it `value`, but I don't actually like that.

Comment (BenTheElder) [reply slack/types.go:110]: Certainly not going to block a PR by bokeshedding names, but this one does seem silly.

**Pull request merged**